src/k2/persistence/logStream/LogStream.h

```diff
@@ -0,0 +1,163 @@
+}
+
+namespace log {
+inline thread_local k2::logging::Logger logstream("k2::logstream");
+}
+
+class LogStreamBase;
+class LogStream;
+class MetadataMgr;
+
+// a base class that provide operations to handle the plogs. it will be used by both log stream and metadata manager
+class LogStreamBase{
+private:
+    // to store the information of each plog id used by the log stream
+    struct PlogInfo {
+        uint32_t currentOffset;
+        bool sealed;
+        uint32_t index;
+    };
+
+    struct MetadataElement{
+        String plogId;
+        uint32_t start_offset;
+        uint32_t size;
+    };
+public:
+    LogStreamBase();
+    ~LogStreamBase();
+    
+    // create a log stream base
+    seastar::future<> create();
+
+    // write data to the the current plog, return the latest Plog ID and latest offset
+    seastar::future<std::pair<String, uint32_t> > append(Payload payload);
+
+    // write data to the the current plog with the start offset, return the latest Plog ID and latest offset
+    seastar::future<std::pair<String, uint32_t> > append(Payload payload, String plogId, uint32_t offset);
+
+    // read all the data from this log stream
+    seastar::future<std::vector<Payload> > read(String start_plogId, uint32_t start_offset, uint32_t size);
+
+    // reload the _usedPlogIdVector and _usedPlogInfo for replay purpose
+    seastar::future<Status> reload(std::vector<dto::MetadataRecord> records);
+
+    // init the plog client client
+    seastar::future<> init_plog_client(String cpo_url, String persistenceClusrerName);
+
+    // obtain the target plog status
+    seastar::future<std::tuple<Status, std::tuple<uint32_t, bool>>> get_plog_status(String plogId);
+private:
+    PlogClient _client;
+    // the maximum size of each plog
+    constexpr static uint32_t PLOG_MAX_SIZE = 16 * 1024 * 1024;
+    // how many redundant plogs it will create in advance 
+    constexpr static uint32_t PLOG_POOL_SIZE = 1;
+
+    // whether this log stream base has been created 
+    bool _create = false; 
+
+    // The vector to store the redundant plog Id
+    std::vector<String> _redundantPlogPool;
+    // The vector to store the used plog Id
+    std::vector<String> _usedPlogIdVector;
+    // The map to store the used plog information
+    std::unordered_map<String, PlogInfo> _usedPlogInfo;
+    // whether the logstream is switching the plog 
+    bool _switched;
+    std::vector<seastar::promise<>> _requestWaiters;
+
+    // a virtual API that used to persist the Plog Id and sealed offset of each used plog
+    virtual seastar::future<Status> _persistSelfMetadata(uint32_t sealed_offset, String new_plogId)=0;
+
+    // when exceed the size limit of current plog, we need to seal the current plog, write the sealed offset to metadata, and write the contents to the new plog
+    seastar::future<std::pair<String, uint32_t> > _switchPlog(Payload payload);
+};
+
+
+class LogStream:public LogStreamBase{
```

Comment:
Do we need inheritance?

Reply:
The Logstream and MetadataManager will have many common APIs. We could test the performance damage of this inheritance later